include/deal.II/multigrid/mg_coarse.h

```diff
@@ -284,6 +333,99 @@ MGCoarseGridApplySmoother<VectorType>::operator()(const unsigned int level,
   coarse_smooth->smooth(level, dst, src);
 }
 
+/* ------------------ Functions for MGCoarseGridApplyPreconditioner ----------*/
+
+template <class VectorType, class PreconditionerType>
+MGCoarseGridApplyPreconditioner<VectorType, PreconditionerType>::
+  MGCoarseGridApplyPreconditioner()
+  : preconditioner(0, typeid(*this).name())
+{}
+
+
+
+template <class VectorType, class PreconditionerType>
+MGCoarseGridApplyPreconditioner<VectorType, PreconditionerType>::
+  MGCoarseGridApplyPreconditioner(const PreconditionerType &preconditioner)
+  : preconditioner(&preconditioner, typeid(*this).name())
+{}
+
+
+
+template <class VectorType, class PreconditionerType>
+void
+MGCoarseGridApplyPreconditioner<VectorType, PreconditionerType>::initialize(
+  const PreconditionerType &preconditioner_)
+{
+  preconditioner = &preconditioner_;
+}
+
+
+
+template <class VectorType, class PreconditionerType>
+void
+MGCoarseGridApplyPreconditioner<VectorType, PreconditionerType>::clear()
+{
+  preconditioner = 0;
+}
+
+
+namespace internal
+{
+  namespace MGCoarseGridApplyPreconditioner
+  {
+    template <class VectorType,
+              class PreconditionerType,
+              typename std::enable_if<
+                std::is_same<typename VectorType::value_type, double>::value,
+                VectorType>::type * = nullptr>
+    void
+    solve(const PreconditionerType preconditioner,
+          VectorType &             dst,
+          const VectorType &       src)
+    {
+      // to allow the case that the preconditioner was only set up on a
+      // subset of processes
+      if (preconditioner != nullptr)
+        preconditioner->vmult(dst, src);
+    }
+
+    template <class VectorType,
+              class PreconditionerType,
+              typename std::enable_if<
+                !std::is_same<typename VectorType::value_type, double>::value,
+                VectorType>::type * = nullptr>
+    void
+    solve(const PreconditionerType preconditioner,
+          VectorType &             dst,
+          const VectorType &       src)
+    {
+      LinearAlgebra::distributed::Vector<double> src_;
+      LinearAlgebra::distributed::Vector<double> dst_;
+
+      src_ = src;
+      dst_ = dst;
+
+      // to allow the case that the preconditioner was only set up on a
+      // subset of processes
+      if (preconditioner != nullptr)
+        preconditioner->vmult(dst_, src_);
```

Comment:
I think this hard-codes the code to `LA:d:Vector`, which will not work with the PETSc-based preconditioners at the moment. I guess it is a bit out-of-scope right now, but I think we should work towards suitable interfaces also for that latter case. Having written https://github.com/exadg/exadg/blob/b69336a4496096f57b18d32deefbffcfe9028f5f/include/exadg/solvers_and_preconditioners/preconditioners/preconditioner_amg.h#L235-L253 and related functionality in an external code, I think something like that in deal.II would be great (but not for this PR, clearly).